src/editor.js

```diff
@@ -460,115 +504,99 @@ export class OdooEditor extends EventTarget {
-                    const record = result[residx];
-                    this._collaborativeLastSynchronisedId = record.id;
-                    if (
-                        index < this._historySteps.length &&
-                        record.id === this._historySteps[index].id
-                    ) {
-                        index++;
-                        continue;
-                    }
-                    updated = true;
-
-                    // we are not synched with the server anymore, rollback and replay
-                    while (this._historySteps.length > index) {
-                        this.historyRollback();
-                        this._historySteps.pop();
-                    }
-
-                    if (record.id === 1) {
-                        this.editable.innerHTML = '';
-                    }
-                    this.historyApply(record.mutations);
-
-                    record.mutations = record.id === 1 ? [] : record.mutations;
-                    this._historySteps.push(record);
-                    index++;
-                }
-                if (updated) {
-                    this._historySteps.push({
-                        cursor: {},
-                        mutations: [],
-                    });
-                }
-                this.observerActive();
-                this.historyFetch();
-            })
-            .catch(() => {
-                // TODO: change that. currently: if error on fetch, fault back to non collaborative mode.
-                this._isCollaborativeActive = false;
+        this.observerUnactive();
+        const localStep = this._historySteps[newStep.index];
+        if (!localStep && newStep.index > this._historySteps.length) {
+            this.options.collaborative.requestSynchronization();
+        } else if (!localStep) {
+            // apply step
+            this.historyApply(newStep.mutations);
+            this._historySteps.push(newStep);
+        } else if (localStep.id !== newStep.id) {
+            this._computeHistoryCursor();
+            //rollback and apply
+            const stepsToReapply = [];
+            while (this._historySteps.length - 1 >= newStep.index) {
+                const poppedStep = this._historySteps.pop();
+                this.historyRevert(poppedStep);
+                // Put the step at the beginning of the array, so that the first
+                // reverted step is the last reapplied
+                stepsToReapply.unshift(poppedStep);
+            }
+            this.historyApply(newStep.mutations);
+            this._historySteps.push(newStep);
+            stepsToReapply.forEach(step => {
+                this.historyApply(step.mutations);
+                this._historySteps.push({ ...step, index: step.index + 1 });
             });
+            this.historySetCursor(this.historyGetCurrentStep());
+        }
+        //else if (localStep && newStep.id === localStep.id)
```

Comment:
Remove commented line.

Reply:
I'd rather not, it's there to document the explicit lack of action for that scenario. At first I left the actual code with the `// do nothin'` comment, but I felt it was a little confusing to actually test and branch on a condition where we do nothing.

Reply:
Alors mettre un commentaire au dessus pour expliquer pourquoi on ne gère pas ce cas là. Histoire qu'on comprenne bien que c'est pas une ligne oubliée quoi parce que c'est à ça que ça ressemble pour l'instant... ^^'

Reply:
C'est expliqué ligne 548.